src/Data/Rpm/DbHeaderBlob.hs

```diff
@@ -0,0 +1,490 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module Data.Rpm.DbHeaderBlob (
+  readPackageInfo,
+  PkgInfo (..),
+  -- everything below here is exposed for testing
+  HeaderBlob (..),
+  EntryInfo (..),
+  IndexEntry (..),
+  RegionInfo (..),
+  headerBlobInit,
+  hdrblobVerifyRegion,
+  emptyRegionInfo,
+  calcDataLength,
+  -- consts
+  rpmTagHeaderImg,
+  rpmTagHeaderImmutable,
+  rpmTagHeaderSignatures,
+  rpmStringType,
+  regionTagType,
+  regionTagCount,
+  hdrblobImport,
+  rpmStringArrayType,
+  rpmI18NstringType,
+  rpmInt64Type,
+  rpmInt32Type,
+) where
+
+import Control.Applicative (liftA2)
+import Control.Monad (foldM, join, replicateM, unless, when)
+import Data.Bifunctor (bimap, first)
+import Data.Binary.Get (ByteOffset, Get, getInt32be, getWord32be, label, runGetOrFail)
+import Data.Bits ((.&.))
+import Data.ByteString qualified as BS
+import Data.ByteString.Lazy qualified as BLS
+import Data.Char (ord)
+import Data.Int (Int32)
+import Data.IntMap qualified as Map
+import Data.List.NonEmpty (NonEmpty ((:|)))
+import Data.List.NonEmpty qualified as NonEmpty
+import Data.String.Conversion (decodeUtf8)
+import Data.Text (Text)
+import Data.Vector qualified as V
+import Data.Word (Word32, Word8)
+import Text.Printf (printf)
+
+-- Constants from
+-- https://github.com/knqyf263/go-rpmdb/blob/master/pkg/rpmtags.go#L3
+rpmTagHeaderImmutable :: Int32
+rpmTagHeaderImmutable = 63
+
+rpmTagHeaderSignatures :: Int32
+rpmTagHeaderSignatures = 62
+
+rpmTagHeaderImg :: Int32
+rpmTagHeaderImg = 61
+
+rpmTagHeaderI18nTable :: Int32
+rpmTagHeaderI18nTable = 100
+
+rpmStringType :: Word32
+rpmStringType = 6
+
+rpmStringArrayType :: Word32
+rpmStringArrayType = 8
+
+rpmI18NstringType :: Word32
+rpmI18NstringType = 9
+
+rpmInt64Type :: Word32
+rpmInt64Type = 5
+
+rpmInt32Type :: Word32
+rpmInt32Type = 4
+
+-- https://github.com/knqyf263/go-rpmdb/blob/master/pkg/entry.go#L71
+-- Size in bytes.
+entryInfoSize :: Int32
+entryInfoSize = 16
+
+-- https://github.com/knqyf263/go-rpmdb/blob/master/pkg/entry.go#L18
+headerMaxBytes :: Int32
+headerMaxBytes = 256 * 1024 * 1024
+
+-- https://github.com/knqyf263/go-rpmdb/blob/master/pkg/entry.go#L14
+regionTagCount :: Int32
+regionTagCount = entryInfoSize
+
+-- https://github.com/knqyf263/go-rpmdb/blob/master/pkg/entry.go#L15
+regionTagType :: Word32
+regionTagType = 7
+
+typeSizes :: V.Vector Int32
+typeSizes =
+  V.fromList
+    [ 0 -- RPM_NULL_TYPE
+    , 1 -- RPM_CHAR_TYPE
+    , 1 -- RPM_INT8_TYPE
+    , 2 -- RPM_INT16_TYPE
+    , 4 -- RPM_INT32_TYPE
+    , 8 -- RPM_INT64_TYPE
+    , -1 -- RPM_STRING_TYPE
+    , 1 -- RPM_BIN_TYPE
+    , -1 -- RPM_STRING_ARRAY_TYPE
+    , -1 -- RPM_I18NSTRING_TYPE
+    , 0
+    , 0
+    , 0
+    , 0
+    , 0
+    , 0
+    ]
```

Comment:
The main reason that I copied this into a vector is that the indices for each size correspond to the type of the tag, which appears directly in the blob. I felt that making a full type might be overkill because data of the custom type would be really ephemeral and sit in the middle of a fairly straightforward mapping of number -> number. I can see how the types are better for documentation purposes though so I will add it.